### ISSUE TITLE
Incorrect Report that an Object may be null

### ISSUE THREAD

**Issue by legodude17** [microsoft/TypeScript#43537]
# Bug Report

### 🔎 Search Terms

 - `Object is possibly 'null'`
 - `undefined`

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about unexpected warnings about possible 'null's and 'undefined's.

I tried versions `3.3.3333` and `4.3.0-beta`, though I did not try any in the middle.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210405#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBBCcQQ0JzJAqgADesEJTm2+SZOgY0HYkA5nHYiGkPwAvqAAD6DKLQEUAbQAuqAALygWXM1mo1AcxmgLk8vmgABMIpF8oA3MFEaAABR0wQBSB6Mh4a22yDJACUyoAfKA6QEWWQ2TQAIQqgXQaAer2MxQlEp22UABnlAY12wCet5ytAAEYLSLgiocA0PGQvIh4AMGCD2DRvm4zMx2AxmJB+LBoOpEjX4HoRF4AEQAUVWDEQNX0g6SOMH4AcqAA5MIAHKQCSDoIqADy3zIu0E7a0kGOqEcClc0jwfsglZ0oyw+hQCEQMOEVwbEir0REBg4OO6Fg2C4bgghKIA)

### 💻 Code

```ts
const refs: {current: {value: number} | null}[] = [{current: { value: 2}}];

if (refs.every((ref) => ref.current != null)) {
    refs[0].current.value = 1;
}
```

### 🙁 Actual behavior

"Object is possibly 'null'" warning on `refs[0].current`

### 🙂 Expected behavior

No warnings, as they cannot be null since the `refs.every(...)` is checking that.

**Comment by DanielRosenwasser** [comment]
Similar to https://github.com/microsoft/TypeScript/issues/16069, maybe https://github.com/microsoft/TypeScript/issues/18562.

One of the problems with this is that if something mutates the underlying array, your assumptions would be invalidated.

**Comment by andrewbranch** [comment]
I think we can call it a duplicate of / encompassed by https://github.com/microsoft/TypeScript/issues/38390

**Comment by legodude17** [comment]
Yes we can!
